src/index.rs

```diff
@@ -1,65 +1,116 @@
 use anyhow::anyhow;
-use canonical_path::{CanonicalPath, CanonicalPathBuf};
-use itertools::Itertools;
+use log;
+use serde::{Deserialize, Serialize};
+use serde_with::serde_as;
 use std::collections::{HashMap, HashSet};
 use std::fs::{self, File, Metadata};
-use std::io::{BufRead, BufReader, Write};
+use std::io::BufRead;
+use std::io::BufReader;
+use std::io::Write;
 use std::ops::Add;
 use std::path::{Path, PathBuf};
 use std::str::FromStr;
-use std::time::{Duration, SystemTime, UNIX_EPOCH};
+use std::time::UNIX_EPOCH;
+use std::time::{Duration, SystemTime};
 use walkdir::{DirEntry, WalkDir};
 
-use log;
+use crate::{id::ResourceId, ArklibError, Result, ARK_FOLDER, INDEX_PATH};
 
-use crate::id::ResourceId;
-use crate::{ArklibError, Result, ARK_FOLDER, INDEX_PATH};
+pub const RESOURCE_UPDATED_THRESHOLD: Duration = Duration::from_millis(1);
+pub type Paths = HashSet<PathBuf>;
 
-#[derive(Eq, Ord, PartialEq, PartialOrd, Hash, Clone, Debug)]
+/// IndexEntry represents a [`ResourceId`] and the time it was last modified
+#[derive(
+    Eq, Ord, PartialEq, PartialOrd, Hash, Clone, Debug, Serialize, Deserialize,
+)]
 pub struct IndexEntry {
+    /// The time the resource was last modified
     pub modified: SystemTime,
+    /// The resource's ID
     pub id: ResourceId,
 }
 
-#[derive(PartialEq, Clone, Debug)]
+/// Represents an index of resources stored as files
+/// in the filesystem under some prefix, or "root".
+///
+/// This struct maintains a mapping from resource IDs to their corresponding
+/// file paths, as well as the mapping of the opposite direction from file
+/// paths to index entries, which are simply resource IDs with modification
+/// timestamps.
+///
+/// The paths stored in the mappings are not absolute paths, but are relative
+/// to the "root". See methods `TODO` and `TODO` in the API, which can provide
+/// the end users with absolute/canonical(TODO) paths for the convenince.
+/// Otherwise, the end user is responsible for managing the working dir
+/// and relative paths resolution.
```

Comment:
In the end, our API should consist of 2 halves:
* "Snapshot API" allowing to query particular paths or ids, separate functions for relative and absolute paths. We could also implement something like Iter interface.
* "Reactive API" allowing updates handling without explicit queries. It can be done in pull-based manner like now (update_all), push-based using Tokio streams and/or registering handlers which would be called automatically (variant of push-based) 

We should discuss and create new issues for this.

Reply:
Opened a ticket #91 to further discuss this